python/ray/util/client/dataclient.py

```diff
@@ -119,21 +193,25 @@ def _blocking_send(self, req: ray_client_pb2.DataRequest
             req_id = self._next_id()
             req.req_id = req_id
             self.request_queue.put(req)
-        data = None
+            self.outstanding_requests[req_id] = req
         with self.cv:
             self.cv.wait_for(
                 lambda: req_id in self.ready_data or self._in_shutdown)
-            if self._in_shutdown:
-                from ray.util import disconnect
-                disconnect()
-                raise ConnectionError(
-                    "Sending request failed because the data channel "
-                    "terminated. This is usually due to an error "
-                    f"in handling the most recent request: {req}. Ray Client "
-                    "has been disconnected.")
-            data = self.ready_data[req_id]
-            del self.ready_data[req_id]
-        return data
+            if not self._in_shutdown:
+                data = self.ready_data[req_id]
+                with self.lock:
+                    del self.outstanding_requests[req_id]
+                    self._acknowledge(req_id)
+                del self.ready_data[req_id]
+                return data
+        # Handle shutdown after dropping the condvar
+        from ray.util import disconnect
+        disconnect()
```

Comment:
Just a note: currently `_blocking_send()` is the only function where Ray client disconnects after shutdown. `disconnect()` is not called after `self._in_shutdown` in the `_async_send()` path. This is fixed in #18539 and we would need to merge the changes.

Reply:
Sounds good, 18539 has been merged into master + this PR